modules/react/common/lib/utils/components.ts

```diff
@@ -78,26 +76,43 @@ type ExtractHTMLAttributes<
+  ? [TElement] extends [never] // test if user passed `never` for the `TElement` override. We have to test `never` first, otherwise TS gets confused and `ExtractProps` will return `never`. https://github.com/microsoft/TypeScript/issues/23182
+    ? P // if `TElement` was `never`, return only the inferred props `P`
+    : TElement extends undefined // else test if TElement was defined
+    ? E extends keyof JSX.IntrinsicElements // test if the inferred element `E` is in `JSX.IntrinsicElements`
+      ? P & ExtractHTMLAttributes<JSX.IntrinsicElements[E]> // `TElement` wasn't explicitly defined, so let's fall back to the inferred element's HTML attribute interface + props `P`
+      : P & ExtractPropsFromComponent<E> // E isn't in `JSX.IntrinsicElements`, return inferred props `P` + props extracted from component `E`. It would be nice to use `ExtractProps` again here, but that creates a circular dependency
+    : TElement extends keyof JSX.IntrinsicElements // `TElement` was defined, test if it is in `JSX.IntrinsicElements`
+    ? P & ExtractHTMLAttributes<JSX.IntrinsicElements[TElement]> // `TElement` is in `JSX.IntrinsicElements`, return inferred props `P` + HTML attributes of `TElement`
+    : P & ExtractPropsFromComponent<TElement> // `TElement` is not in `JSX.IntrinsicElements`, return inferred props `P` + props extracted from component `TElement`. It would be nice to use `ExtractProps` again here, but that creates a circular dependency
+  : ExtractPropsFromComponent<TComponent>; // `TComponent` does not extend `ElementComponent`. Return props extracted from component `TComponent`
+
+// Extract props from a component. This type is only necessary because `ExtractProps` cannot
+// reference itself, creating a circular dependency. Instead, we define this type to allow for at
+// least 1 level of nesting.
+type ExtractPropsFromComponent<TComponent> = TComponent extends ElementComponent<infer E, infer P> // test if `TComponent` is an `ElementComponent`, while inferring both default element and props associated
+  ? E extends keyof JSX.IntrinsicElements // test if the inferred element `E` is in `JSX.IntrinsicElements`
+    ? P & ExtractHTMLAttributes<JSX.IntrinsicElements[E]> // return inferred props `P` + HTML attributes of inferred element `E`
+    : P // `E` wasn't a key of `JSX.IntrinsicElements`, so just return the inferred props `P`
   : TComponent extends Component<infer P> // test if `TComponent` is a `Component`, while inferring props `P`
   ? P // it was a `Component`, return inferred props `P`
-  : {}; // `TComponent` does not extend either `ElementComponent` or `Component`, return an empty object
+  : TComponent extends React.ComponentType<infer P> // test if `TComponent` is a `React.ComponentType` (class or functional component)
+  ? P // it was a `React.ComponentType`, return inferred props `P`
+  : {}; // We don't know what `TComponent` was, return an empty object
```

Comment:
Complete refactor of this type. I've added some new features while fixing the bug:
- `ExtractProps` now supports component overrides: `ExtractType<ElementComponent<'div', Props>, typeof SomeComponent`. We support `as` for other components, `ExtractProps` should as well

Reply:
I ran into this issue when refactoring: https://www.typescriptlang.org/play?#code/C4TwDgpgBAKhDOwbggHgFBVlCAPYEAdgCbxQCuJEAZgJaETFQA+UDAbhAE5QC8FVOg2LoAfH2x4CJMpWI16jKAHplOADbxoBRFFrUoAAxiGoAdwCGZeUMaYoAfigBvAIwAuKIi70A5gF8sVSMTcytCAHJgHFwwdVoAY1pgdRAoG0URLE9nACZPbz9AlTVjU0syQw5uQwAaKB9fAAtgJxgmi2AIsmAm6AB7QlSoXr8odRpgADoZqABRTkJonWAxvDjE5OHqfp4qiE4uUz6uaGJ+hEjos12AaymoAElorQgAWx7+qF8IaPV+xAzKbodCgSBQABi-X6rgkcEQyHBwTcBWAPkIAXqCV2pwSwFBKEh0NycIQSBQqGqXHEwSp9XgTX65HUTAARtA8qj0QEQcF4eTwRZfBZ6PVWeRrskmlALIQYhsktF5AQ8bRBlB+gZ9odTAB9dk7U66oxyBTCQwE8H8xEQXIYLAwGLSUgCDLCFhsA7cCSm2wicT8ADaMAAuk6iC7A1Sw8EIJptGS9FrQhUjFTDA8AMoQaAtYBgeDuVS+KXkVlTbFvZRvRJcAGa4DKG2ZhI+MCN2jweDkBDKXIAZlcAA5cvYnJyvGiiiUQuUrGmvUd7DkPJPuYFgmUoG9yLp2SbBJkLZboAAhCxcWH8a0oEnI1eFDH+LE4iB4k9Qc9cEnXsk2u1UjSagTo+mJQNiXC4sAQA